File: Extended/Erlang/url_shortener.erl
-module(url_shortener).
-export([start/0, shorten/2, get_url/2, generate_short_code/1]).

% Start a new URL shortener process
start() ->
    spawn(fun() -> loop(#{}) end).

% Shorten a URL
shorten(Pid, Url) ->
    Pid ! {shorten, Url, self()},
    receive
        {short_code, ShortCode} -> ShortCode
    after 5000 ->
        timeout
    end.

% Get original URL from short code
get_url(Pid, ShortCode) ->
    Pid ! {get_url, ShortCode, self()},
    receive
        {url, Url} -> Url;
        not_found -> not_found
    after 5000 ->
        timeout
    end.

% Generate random short code
generate_short_code(Length) ->
    Characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789",
    generate_short_code(Length, Characters, []).

generate_short_code(0, _Characters, Acc) ->
    lists:reverse(Acc);
generate_short_code(Length, Characters, Acc) ->
    Index = rand:uniform(length(Characters)),
    Char = lists:nth(Index, Characters),
    generate_short_code(Length - 1, Characters, [Char | Acc]).

% Main process loop
loop(UrlMap) ->
    receive
        {shorten, Url, From} ->
            ShortCode = generate_short_code(6),
            NewUrlMap = maps:put(ShortCode, Url, UrlMap),
            From ! {short_code, ShortCode},
            loop(NewUrlMap);
        {get_url, ShortCode, From} ->
            case maps:find(ShortCode, UrlMap) of
                {ok, Url} ->
                    From ! {url, Url};
                error ->
                    From ! not_found
            end,
            loop(UrlMap)
    end.

% Example usage
example() ->
    Shortener = start(),
    ShortCode = shorten(Shortener, "https://www.example.com"),
    io:format("Short URL: ~s~n", [ShortCode]),
    case get_url(Shortener, ShortCode) of
        not_found ->
            io:format("URL not found~n");
        Url ->
            io:format("Original URL: ~s~n", [Url])
    end.

File: Extended/OCaml/url_shortener.ml
module URLShortener = struct
  type t = (string, string) Hashtbl.t
  
  let create () = Hashtbl.create 16
  
  let generate_short_code length =
    let characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789" in
    let char_count = String.length characters in
    let result = Bytes.create length in
    for i = 0 to length - 1 do
      let index = Random.int char_count in
      Bytes.set result i characters.[index]
    done;
    Bytes.to_string result
  
  let shorten shortener url =
    let short = generate_short_code 6 in
    Hashtbl.add shortener short url;
    short
  
  let get_url shortener short =
    try Some (Hashtbl.find shortener short)
    with Not_found -> None
end

(* Example usage *)
let () =
  Random.self_init ();
  let shortener = URLShortener.create () in
  let short = URLShortener.shorten shortener "https://www.example.com" in
  Printf.printf "Short URL: %s\n" short;
  match URLShortener.get_url shortener short with
  | Some original -> Printf.printf "Original URL: %s\n" original
  | None -> Printf.printf "URL not found\n"
